Реализован базовый функционал Hacker News

JstUsername opened the pull request:
https://stackblitz.com/edit/vitejs-vite-vjmdmv

JstUsername committed:
feat(.gitignore): добавлен .gitignore

JstUsername committed:
feat(dependencies): добавлены необходимые для проекта зависимости

JstUsername committed:
feat(vite): добавлены файлы конфигурации для vite

JstUsername committed:
feat(ts): добавлены файлы конфигурации для typescript

JstUsername committed:
feat(eslint): добавлены файлы конфигурации для eslint

JstUsername committed:
feat(prettier): добавлены файлы конфигурации для prettier

JstUsername committed:
fix(dependencies): из зависимостей удален ant, добавлен styled-components

JstUsername committed:
feat(.idea): инициализирован WebStorm

JstUsername committed:
feat(svgr): в проект добавлен vite-plugin-svgr

JstUsername committed:
feat(favicon): добавлен и подключен полный набор фавиконов

JstUsername committed:
feat(logo.svg): в /assets добавлен логотип

JstUsername committed:
feat(source-code-pro): в зависимости добавлен шрифт source-code-pro

JstUsername committed:
feat(hour-glass): добавлена иконка для лоадера

JstUsername committed:
feat(logo-cropped): добавлен компактный вариант логотипа

JstUsername committed:
feat(theme): добавлена тема для styled-components

JstUsername committed:
feat(Header): добавлен header и его стили

JstUsername committed:
feat(NewsListItem): добавлен компонент новости, его типы и стили

JstUsername committed:
feat(NewsList): добавлен компонент списка новостей и его стили

JstUsername committed:
feat(store): добавлен store для zustand, а также типы для стейта

JstUsername committed:
feat(App): добавлен компонент App

JstUsername committed:
feat(Main): добавлен компонент Main

JstUsername committed:
fix(theme): добален backgroundLightTransparent, исправлено использование темы в скроллбаре

JstUsername committed:
fix(NewsListWrapper): добавлено размещение 3-х новостей в строку, на больших экранах

JstUsername committed:
fix(NewsListItem): цвет фона карточек новостей  заменен на backgroundLightTransparent

JstUsername committed:
fix(backgroundLight): backgroundLight переименован в background, добавлена прозрачность

JstUsername committed:
fix(Header): Link переименован в StyledLink

JstUsername committed:
feat(Footer): добавлен компонент Footer, его типы и стили

JstUsername committed:
fix(updateNewsList): запросы выполняются последовательно, для ускорения загрузки

JstUsername committed:
fix(NewsTitle): добавлен word-break для NewsTitle

JstUsername committed:
fix(GlobalStyle): удалено свойство height для body

JstUsername committed:
fix(Footer): исправлено позиционирование футера.

Позиционирование больше не зависит от isLoading, а реализовано с помощью медиа-запросов в родительском компоненте.

JstUsername committed:
feat(NotFoundPage): добавлен компонент NotFoundPage и его стили

JstUsername committed:
feat(Layout): Header и Footer вынесены в Layout

JstUsername committed:
feat(HomePage): NewsList вынесен в компонент HomePage

JstUsername committed:
fix(App): реализован роутинг с помощью React Router

JstUsername committed:
feat(updateNewsList): добавлена функция для обновления списка последних новостей

JstUsername committed:
fix(updateNewsList): по нажатию на кнопку обновляются последние новости, а не весь список

JstUsername committed:
feat(updateNewsList): добавлен useEffect для обновления списка последних новостей раз в минуту

JstUsername committed:
feat(InternalServerError): добавлен компонент для 500 ошибки

JstUsername committed:
feat(isServerDown): в состояние добавлен isServerDown, а также обработка ошибки 500

JstUsername committed:
feat(InternalServerError): в NewsList добавлен условный рендеринг InternalServerError

JstUsername committed:
feat(useNavigate): по нажатию на лого происходит переход домой

JstUsername committed:
fix(Header.styled): для лого добавлен cursor pointer, исправлены padding

JstUsername committed:
fix(NewsList.styled): исправлены padding

JstUsername committed:
feat(/news/:id): добавлен динамический адрес для страницы новости

JstUsername committed:
fix(isLoading, isServerDown): переименованы в newsLoading, newsServerDown

fix(isLoading, isServerDown): переименованы в newsLoading, newsServerDown

fix(isLoading, isServerDown): переименованы в newsLoading, newsServerDown

fix(isLoading, isServerDown): переименованы в newsLoading, newsServerDown

fix(isLoading, isServerDown): переименованы в newsLoading, newsServerDown

JstUsername committed:
fix(NewsListType): скорректированы типы, согласно рекомендациям API

JstUsername committed:
fix(useEffect): запрос данных выполняется только при первой загрузке списка новостей

JstUsername committed:
feat(body2): в тему добавлен вариант body2 для текста

JstUsername committed:
fix(NewsListItemProps): скорректированы типы, согласно рекомендациям API

JstUsername committed:
feat(NewsItemPage): добавлен компонент отображающий конкретную новость

JstUsername committed:
feat(useNewsItemState): добавлен стейт и его типы для NewsItemPage

fix(store): удалена лишняя строка в store

JstUsername committed:
feat(onClick): добавлен роутинг по нажатию на карточку новости

JstUsername committed:
fix(NewsList, NewsItemPage): тернарные операторы переписаны на if-else, реализован ранний выход

JstUsername committed:
feat(pixel-arrow): добавлена иконка стрелки

JstUsername committed:
feat(CommentsItem): добавлен компонент для отрисовки комментариев, его типы и стили

JstUsername committed:
fix(NewsItemPage): CommentsItem добавлен в NewsItemPage, стили вынесены в отдельный файл

JstUsername committed:
fix(store): добавлены функции, реализующие логику секции комментариев. Обновлены типы store.

Добавлены функции getNewsContent, getNewsItem, addVisibleField, clearingDeleted, setExpandVisible

JstUsername committed:
feat(updateCurrentNews): в зависимости от пути, в header, разные функции обновления

JstUsername committed:
fix(CroppedHNLogo): для компактного логотипа добавлен роутинг домой по нажатию

fix(CroppedHNLogo): для компактного логотипа добавлен роутинг домой по нажатию

IvanKoigerov commented on src/components/CommentsItem/CommentsItem.tsx:
```diff
@@ -0,0 +1,45 @@
+} from './CommentsItem.styled.ts';
+
+export default function CommentsItem({ comment }: CommentsListProps) {
+  const { newsItem, setExpandVisible } = useNewsItemState();
```
На лекции говорил, что так нельзя)

IvanKoigerov commented on src/theme.ts:
```diff
@@ -0,0 +1,72 @@
+  },
+  svg: {
+    //prettier-ignore
+    background: '"data:image/svg+xml,%3Csvg xmlns=\'http://www.w3.org/2000/svg\' width=\'400\' height=\'400\' viewBox=\'0 0 800 800\'%3E%3Cg fill=\'none\' stroke=\'%23022741\' stroke-width=\'1\'%3E%3Cpath d=\'M769 229L1037 260.9M927 880L731 737 520 660 309 538 40 599 295 764 126.5 879.5 40 599-197 493 102 382-31 229 126.5 79.5-69-63\'/%3E%3Cpath d=\'M-31 229L237 261 390 382 603 493 308.5 537.5 10 ...
```
не такую тему лучше в тему не закидывать)

JstUsername committed:
fix(svg): фоновый svg напрямую передан в background-image

JstUsername committed:
fix(body): вариант body переименован в body1 и заменен во всех файлах

JstUsername committed:
fix(store): исправлено использование переменных и функций из стейта

Kimiyori commented on src/components/CommentsItem/CommentsItem.styled.ts:
```diff
@@ -0,0 +1,53 @@
+import styled from 'styled-components';
+import { CommentsChildItemWrapperProps, ExpandIconProps } from './CommentsItem.types.ts';
+import ArrowIcon from '../../assets/pixel-arrow.svg?react';
```
лучше на абсолютные пути переписать наверное

IvanKoigerov replied:
Не парься в стажерских тасках это не нужно

Kimiyori commented on src/components/Header/Header.tsx:
```diff
@@ -0,0 +1,28 @@
+  return (
+    <StyledHeader>
+      <CroppedHNLogo title="Logo" onClick={() => navigate('/')} />
+      <HNLogo title="Logo" onClick={() => navigate('/')} />
```
почему не в <a> тага?

IvanKoigerov replied:
Почему не тег А 
Правильно, нужно в таких местах использовать ашки потому что это линк
То есть на нее можно правой кнопкой мыши кликать и открывать в новой вкладке
Или средней и тд

Kimiyori commented on src/components/NewsList/NewsList.tsx:
```diff
@@ -0,0 +1,53 @@
+
+  useEffect(() => {
+    newsList.length === 0 ? getNewsList() : null;
+  }, [newsList, getNewsList]);
```
а в рекурсию не упадет?

Kimiyori commented on src/components/NewsList/NewsList.tsx:
```diff
@@ -0,0 +1,53 @@
+          title={newsItem.title}
+          points={newsItem.points}
+          user={newsItem.user}
+          time_ago={newsItem.time_ago}
```
{...newsItem}

Kimiyori commented on src/components/NewsList/NewsList.tsx:
```diff
@@ -0,0 +1,53 @@
+  }, [updateNewsList]);
+
+  if (newsServerDown) {
+    return <InternalServerError />;
```
не хочешь потрогать suspense? от таких костылей избавились бы

Kimiyori commented on src/pages/NewsItemPage/NewsItemPage.tsx:
```diff
@@ -0,0 +1,114 @@
+  const timestampToDate = (timestamp: number) => {
+    const date = new Date(timestamp * 1000);
+    const year = date.getFullYear() - 2000;
+    const month = (date.getMonth() + 1).toString().padStart(2, '0');
+    const day = date.getDate().toString().padStart(2, '0');
+    const hours = date.getHours().toString().padStart(2, '0');
+    const minutes = date.getMinutes().toString().padStart(2, '0');
```
мб в функцию pad вынести?

JstUsername replied:
Функция переписана с применением date-fns.

Kimiyori commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+    !auto ? set({ newsLoading: true }) : null;
+    const response = await fetch(`${newsListUrl[0]}?t=${new Date().getTime()}`);
+    if (response.status === 500) {
+      !auto ? set({ newsLoading: false }) : null;
```
&&

Kimiyori commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+      return;
+    }
+    let fetchPromise = await response.json();
+    const newData = fetchPromise.filter((newNewsItem: NewsListType) =>
+      get().newsList.every((oldNewsItem) => newNewsItem.id !== oldNewsItem.id),
```
я до конца не понял, а в чем смысл?) у этой штуки сложность O(n*m), но зачем?

JstUsername replied:
Чуть ниже описал смысл и почему используется при обновлении только один запрос.

IvanKoigerov replied:
по настощему будто бы дешевле было бы просто получать все и отрисовывать, педставь сколько у тебя по итогу происходит пробегов по целому массиву

Kimiyori commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+      !auto ? set({ newsLoading: false }) : null;
+      return;
+    }
+    set({ newsList: newData.concat(get().newsList).slice(0, 100) });
```
а, у тебя же 100 итемов сплошным списком идут... Может лучше пагинацию добавить?

JstUsername replied:
Условие задание такое:
"Показывает последние 100 новостей в виде списка, отсортированного по дате, самые свежие сверху".
https://github.com/web-bee-ru/ru-test-assignments/blob/main/react/01-news-list.md

IvanKoigerov replied:
Ну да, давай оставим эту сотню, тут все корректно
А пагинация норм тема, но сейчас это будет дольше делать

Kimiyori commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+  },
+  updateNewsList: async (auto) => {
+    !auto ? set({ newsLoading: true }) : null;
+    const response = await fetch(`${newsListUrl[0]}?t=${new Date().getTime()}`);
```
почему обновляется только первая часть списка?

JstUsername replied:
Это сделано специально, чтобы каждый раз не кидать по 4 запроса.
Логика такая:
У нас в любом случае раз в минуту происходит автоматическое обновление последних новостей, причем при автоматическом обновлении идет сравнение id'шников и новые новости добавляются только если их не существует в текущем стейте. Это еще пояснение к твоему комментарию чуть выше, где пишешь про сложность O(n*m).
И при ручном обновлении соответственно нам тоже достаточно сделать один запрос, и выделить те новости, которые появились с момента последнего автоматического обновления до момента когда мы вызвали ручное обновление.

Kimiyori replied:
Окей, добавляются новые новости, а старые обновляться не будут? там же могут измениться тоже данные. Твой подход работал бы только в случае с пагинацией, и то надо бы номер страницы еще прокидывать

Kimiyori commented:
словил баг, время неверно отображается
![image](https://github.com/user-attachments/assets/a7f7ae9b-9b7e-42d2-9981-995708e388c8)

Kimiyori commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+      return;
+    }
+    data = await response.json();
+    const addVisibleField = (commentsList: NewsItemType) => {
+      switch (commentsList.level) {
+        case undefined:
+          break;
+        case 0:
+          commentsList.visible = true;
+          break;
+        default:
+          commentsList.visible = false;
+      }
+      commentsList.comments.forEach(addVisibleField);
+    };
```
имеет смысл вынести функцию наверно...и так громоздко выходит

Kimiyori replied:
я честно вообще не понимаю зачем тут стор используется) всмысле это вполне все решается построением дерева и менеджменом состояний видимости комментов в самом компоненте)

Kimiyori replied:
казалось бы функция просто на получение итема новости, а столько внешнй работы)

JstUsername committed:
pr(CroppedHNLogo, HNLogo): логотипы обернуты в тэг Link

JstUsername committed:
pr(NewsList): пропсы для NewsListItem переданы с помощью spread оператора

IvanKoigerov commented on src/components/CommentsItem/CommentsItem.styled.ts:
```diff
@@ -0,0 +1,53 @@
+  ${({ theme }) => theme.typography.body1};
+  color: ${({ theme }) => theme.color.green};
+  p {
+    margin: 12px 0;
+    line-height: 1.2;
+    overflow: hidden;
+    text-overflow: ellipsis;
+  }
+  a {
+    color: ${({ theme }) => theme.color.green};
+    cursor: pointer;
+    white-space: nowrap;
+  }
```
Не круто что ты стили пишешь через теги, хоть и у тебя есть родитель, просто лучше такого избегать

JstUsername replied:
Здесь стилизуется контент, который получается из запроса, не совсем понимаю как по - другому стилизовать.
`<CommentsItemContent dangerouslySetInnerHTML={{ __html: comment.content }} />`

IvanKoigerov commented on src/components/CommentsItem/CommentsItem.styled.ts:
```diff
@@ -0,0 +1,53 @@
+
+export const CommentsChildItemWrapper = styled('div')<CommentsChildItemWrapperProps>`
+  display: ${(props) => (props.$isVisible ? 'block' : 'none')};
+`;
```
А почему не {$isVisible: boolean}? просто, чтобы отдельные сущности не плодить, как тип в данном случае

IvanKoigerov commented on src/components/CommentsItem/CommentsItem.tsx:
```diff
@@ -0,0 +1,46 @@
+        <CommentsItemContent dangerouslySetInnerHTML={{ __html: comment.content }} />
+      </div>
+      {comment.comments.length > 0 && (
+        <>
+          {comment.comments.map((childComment) => (
+            <CommentsChildItemWrapper key={childComment.id} $isVisible={childComment.visible}>
+              <CommentsItem comment={childComment} />
+            </CommentsChildItemWrapper>
+          ))}
+        </>
```
тебе тут не нужны `<></>`

`comment.comments.length ? comment.comments.map : null`
так код будет чище 
Почему тут null? потому что comment.comments.length может быть 0 и тогда он отобразиться на странице

Kimiyori replied:
мне наоборот нравится через &&

IvanKoigerov commented on src/components/Footer/Footer.styled.ts:
```diff
@@ -0,0 +1,44 @@
+  }
+  @media (max-width: ${({ theme }) => theme.breakpoints.sm}) {
+    justify-content: space-between;
+  }
```
обычно межу тем как ты открываешь `{}` не важно где оставляют пустую строку перед и после, для лучшей читабельности

IvanKoigerov commented on src/components/Footer/Footer.styled.ts:
```diff
@@ -0,0 +1,44 @@
+  }
+  bottom: 0;
+  left: 0;
+  right: 0;
```
Не вижу смысла указывать эти значения для футера, растяни его по другому. Просто нет смылса использовать у релетив в принципе эти параметры, просто может быть опасно, что-то где-то может отобразиться не корректно

IvanKoigerov commented on src/components/Header/Header.styled.ts:
```diff
@@ -0,0 +1,69 @@
+  display: block;
+  flex-shrink: 0;
+  cursor: pointer;
+  @media (max-width: ${({ theme }) => theme.breakpoints.sm}) {
+    display: none;
+  }
```
В текущих реалях, лучше всего такие блоки в принципе выпиливать из дома, а не описывать display: none;
Лишний компонент, который никогда не отрисуется на экране, но так как тут ты просто styled используешь то лан пох

IvanKoigerov commented on src/components/Header/Header.styled.ts:
```diff
@@ -0,0 +1,69 @@
+  @media (max-width: ${({ theme }) => theme.breakpoints.sm}) {
+    display: block;
+  }
+`;
```
ну тут тоже самое что и с прошлым

IvanKoigerov commented on src/components/Header/Header.styled.ts:
```diff
@@ -0,0 +1,69 @@
+  }
+`;
+
+export const ButtonLink = styled('button')`
```
ButtonLink кнопки не юзаем для линков

IvanKoigerov replied:
Кстати, не знаю у тебя карточни новостей a или нет, но желательно, чтобы тоже были линками карточки, это важно для конечного удобства пользователя

IvanKoigerov commented on src/components/InternalServerError/InternalServerError.styled.ts:
```diff
@@ -0,0 +1,39 @@
+  }
+  @media (max-width: ${({ theme }) => theme.breakpoints.md}) {
+    height: calc(100vh - (72px + 48px));
+  }
```
Когда такие вещи пишешь, лучше всего описывай в комменте что это такое, что за 88 и 64
чьи это размеры, почему делаешь такой calc

IvanKoigerov replied:
Ну и по возможности таких вещей лучше избегать, никто не знает, когда у тебя может неожиданно измениться размер той же шапки и тд

IvanKoigerov commented on src/components/InternalServerError/InternalServerError.styled.ts:
```diff
@@ -0,0 +1,39 @@
+  ${({ theme }) => theme.typography.h1};
+  font-size: 48px;
+  color: ${({ theme }) => theme.color.purple};
+`;
```
лучше всего для таких вещей по семантике использовать span

IvanKoigerov commented on src/components/InternalServerError/InternalServerError.styled.ts:
```diff
@@ -0,0 +1,39 @@
+  ${({ theme }) => theme.typography.h1};
+  margin: 0;
+  color: ${({ theme }) => theme.color.green};
+`;
```
Лучше всего в одном месте определить h1 с типами ${({ theme }) => theme.typography.h1}; 
и потом уже его в styled закидывать 
Так будет реально проще

IvanKoigerov commented on src/components/InternalServerError/InternalServerError.styled.ts:
```diff
@@ -0,0 +1,39 @@
+  &:hover {
+    color: ${({ theme }) => theme.color.purple};
+  }
+`;
```
В другом файле похожий компонент, лучше все равно их немного по разному называть 
Да и будто бы опять же тут стили повторяются какие-то
Можешь вынести их в какой-то общий абстрактный button

IvanKoigerov commented on src/components/NewsList/NewsList.styled.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+  @media (max-width: ${({ theme }) => theme.breakpoints.md}) {
+    height: calc(100vh - (72px + 48px));
+  }
```
второй раз вижу, вынечи эту вещь опять же в какой-то общий компонент, а тут реюзай

IvanKoigerov commented on src/components/NewsList/NewsList.tsx:
```diff
@@ -0,0 +1,46 @@
+
+  useEffect(() => {
+    newsList.length === 0 ? getNewsList() : null;
+  }, [newsList, getNewsList]);
```
Лучше писать такие вещи не через && или ? : 
А через if, так будет понятнее всего, да и при маунте компонента, корректнее всего было бы перезапрашивать данные под копотом, особенно в ситуации, когда они у тебя так часто меняются

IvanKoigerov commented on src/components/NewsList/NewsList.tsx:
```diff
@@ -0,0 +1,46 @@
+        <NewsListItem key={newsItem.id} {...newsItem} />
+      ))}
+    </NewsListWrapper>
+  );
```
Не круто таки жесткие if писать, лучше всегда такие вещи стараться избегать, тк, у тебя размаунтиться все новости, а потом заново будут маунтиться, из-за этого ты можешь просесть по производительности

IvanKoigerov commented on src/components/NewsList/NewsList.tsx:
```diff
@@ -0,0 +1,46 @@
+
+  if (newsServerDown) {
+    return <InternalServerError />;
+  }
```
ну ошибку норм так выносить, но да Максим правильно говорит, стоит потрогать suspense

IvanKoigerov commented on src/pages/NewsItemPage/NewsItemPage.styled.ts:
```diff
@@ -0,0 +1,136 @@
+  flex-direction: column;
+  justify-content: flex-start;
+  align-items: flex-start;
+  box-sizing: border-box;
```
такие вещи должны быть в css reset или нормалайз, на каждом проекте желательно их использовать

IvanKoigerov commented on src/pages/NewsItemPage/NewsItemPage.styled.ts:
```diff
@@ -0,0 +1,136 @@
+    cursor: pointer;
+    white-space: nowrap;
+  }
+`;
```
ну также теги не круто описывать

IvanKoigerov commented on src/pages/NewsItemPage/NewsItemPage.tsx:
```diff
@@ -0,0 +1,114 @@
+    const hours = date.getHours().toString().padStart(2, '0');
+    const minutes = date.getMinutes().toString().padStart(2, '0');
+    return `${hours}:${minutes} ${day}.${month}.${year}`;
+  };
```
еба дроч, потрогай оч крутую библиотеку date fns

IvanKoigerov commented on src/pages/NewsItemPage/NewsItemPage.tsx:
```diff
@@ -0,0 +1,114 @@
+  const itemLoading = useSelectorItemLoading();
+  const commentsLoading = useSelectorCommentsLoading();
+  const getNewsItem = useSelectorGetNewsItem();
+  const getNewsContent = useSelectorGetNewsContent();
```
Помнишь какое было первое задание (тудушки)? лучше всего подумать как можно распределить компоненты, чтобы избежать таких лютых подписок на все

IvanKoigerov commented on src/pages/NewsItemPage/NewsItemPage.tsx:
```diff
@@ -0,0 +1,114 @@
+  } else {
+    return <NotFoundPage />;
+  }
+}
```
Слишком много вещей делаешь для return, очень нечитабельно выглядит 
Ну и NotFound нужно уж тогда делать раньше 
так же как и лоудер 
Ну и лучше на ошибку сделать один общий компонент, который будет отвечать за отрисовку всех ошибок

IvanKoigerov replied:
Также у тебя повторяется 3 вещи 
Ошибка, лоудер и children, звучит как то, что можно легко вынести в какой-то Layout

IvanKoigerov commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+    set({ newsLoading: true });
+    let data: NewsListType[] | [] = [];
+    for (let i = 0; i < newsListUrl.length; i++) {
+      const response = await fetch(`${newsListUrl[i]}?t=${new Date().getTime()}`);
```
странно, что не используешь Promise.all

IvanKoigerov replied:
он бы в разы упростил бы твой код и сделал бы его более понятным

IvanKoigerov commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+        return;
+      }
+      let fetchPromise = await response.json();
+      i === 3 ? (fetchPromise = fetchPromise.slice(0, 10)) : null;
```
Внутри кода не описывай никогда таким образом работу с проверкой данных, темы нечитабельная
Через if сильно понятнее

IvanKoigerov commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+      return;
+    }
+    set({ newsList: newData.concat(get().newsList).slice(0, 100) });
+    !auto ? set({ newsLoading: false }) : null;
```
много раз повторяешь это слово, корректнее все было бы просто сделать это один раз в начале и один раз в конце

IvanKoigerov commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+    set({ newsList: newData.concat(get().newsList).slice(0, 100) });
+    !auto ? set({ newsLoading: false }) : null;
+  },
+}));
```
Я бы на твоем месте бы по избежании ререндеров, ладно уж, один раз пробежался бы по массиву, если нет новых id то сеть значение, если нет то не сеть, иначе это дороговато таким заниматься на каждый запрос

IvanKoigerov commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+export const useSelectorGetNewsList = () => useNewsListState((state) => state.getNewsList);
+export const useSelectorUpdateNewsList = () => useNewsListState((state) => state.updateNewsList);
+
+const useNewsItemState = create<UseNewsItemType>((set) => ({
```
Нужно бить на два файла, не круто держать два стора в одном файле

IvanKoigerov commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+  commentsLoading: false,
+  itemServerDown: false,
+  getNewsContent: (id, newsList) => {
+    let foundNews = newsList.find((news) => news.id === id);
```
много используешь let, где оно тебе не нужно, не помню какая там структура данных, но как я помню, тебе будто бы не нужна информация о том какой там news из newsList, просто по id получай значение через запрос

IvanKoigerov replied:
Эта штука релеванта, только если ты кешишь, все открытые юзером страницы новостей, без этого просто используем запрос

IvanKoigerov commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+    }
+    data = await response.json();
+    const addVisibleField = (commentsList: NewsItemType) => {
+      switch (commentsList.level) {
```
это визуальная тема, не вижу смысла хранить эту информацию в сторе, согласен с Максимом

IvanKoigerov commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+      set({ newsItem: transformedFoundNews });
+    }
+  },
+  getNewsItem: async (id, newsList, auto) => {
```
очень усложнил эту тему, в этой функции тебе нужно только лишь получать информацию бэка о новости

IvanKoigerov commented on src/store/store.ts:
```diff
@@ -0,0 +1,138 @@
+    set({ newsItem: data });
+    !auto ? set({ itemLoading: false, commentsLoading: false }) : null;
+  },
+  setExpandVisible: (id, newsItem) => {
+    if (newsItem === null) return;
+    const newData = { ...newsItem };
+    const toggleExpandVisible = (obj: NewsItemType) => {
+      if (obj.id === id) {
+        obj.expand = !obj.expand;
+        obj.comments.map((childComments) => {
+          childComments.visible = !childComments.visible;
+        });
+        set({ newsItem: { ...newData } });
+      }
+      if (obj.comments && obj.comments.length > 0) {
+        obj.comments.forEach(toggleExpandVisible);
+      }
+    };
+    toggleExpandVisible(newData);
+  },
```
соответственно лучше это не хранить в сторе

IvanKoigerov commented on src/theme.ts:
Лучше всего делить тему и GlobalStyle
Для разных вещей они и описывать их лучше в разных файлах

JstUsername committed:
pr(date-fns): в проект добавлен date-fns

JstUsername committed:
pr(timestampToDate): конвертация timestamp с помощью date-fns

JstUsername committed:
pr(time_ago): время считается из timestamp, а не берется из запроса

JstUsername committed:
pr(getNewsList, updateNewsList): в getNewsList запросы выполняются параллельно, updateNewsList удален

JstUsername committed:
pr(react-error-boundary): в проект добавлен react-error-boundary

JstUsername committed:
pr(ErrorBoundary): NewsList обернут в ErrorBoundary

JstUsername committed:
pr(Header): нажатию на кнопку в Header выполняет getNewsList()

JstUsername committed:
fix(NewsListWrapper): добавлены стили для минимальной высоты.

Чтобы не "прыгал" футер при загрузке.

JstUsername committed:
pr(getNewsList): добавлен hasMounted для предотвращения двойного выполнения запросов

JstUsername committed:
pr(updateNewsList): updateNewsList заменен на getNewsList

JstUsername committed:
pr(NewsList): добавлен Suspense, пробрасывание ошибки 'Internal Server Error'

JstUsername committed:
fix(NewsList): возвращена проверка длины newsList, для предотвращения лишних запросов

JstUsername committed:
pr(NewsItemPage): добавлен hasMounted для предотвращения двойного выполнения запросов

JstUsername committed:
pr(NewsItemPage): добавлен Suspense, пробрасывание ошибки 'Internal Server Error'

JstUsername committed:
pr(itemPageNotFound): добавлен стейт для 404 ошибки

JstUsername committed:
pr(NewsItemTitle, NewsItemTitleAdditional): вынесены в commons/NewsItemTitle

JstUsername committed:
pr(NewsItemPage): добавлено прокидывание ошибки Page Not Found

pr(NewsItemPage): добавлено прокидывание ошибки Page Not Found

JstUsername committed:
pr(NewsItemPage): блоки контента и комментариев разнесены по отдельным компонентам

pr(NewsItemPage): блоки контента и комментариев разнесены по отдельным компонентам

JstUsername committed:
pr(ErrorHandler): компоненты для ошибок объединены и помещены в ErrorHandler

JstUsername committed:
pr(RefreshButton): StyledButton переименован и вынесен в commons

JstUsername committed:
fix(Error): скорректированы сообщения об ошибках

JstUsername committed:
fix(HomePage): ErrorBoundary удален из HomePage

JstUsername committed:
feat(resetNotFoundPage): в стейт добавлена функция для сброса ошибки 'Page not found'

JstUsername committed:
pr(ErrorEmoji): вместо div использован span для ErrorEmoji

JstUsername committed:
fix(NewsCommentsBlock): исправлен Suspense

JstUsername committed:
pr(App): ErrorBoundary вынесен в App, использован FallbackComponent

JstUsername committed:
pr(CommentsItem): исправлена отрисовка вложенных комментариев

JstUsername committed:
pr(NewsListItemWrapper): карточки новостей стилизованы на основе Link

JstUsername committed:
pr(GlobalStyle): вынесен в src/styles/GlobalStyle.ts

JstUsername committed:
fix(NewsCommentsBlock): удален Suspense, использован условный рендеринг

JstUsername committed:
pr(StyledFooter): исправлено позиционирование StyledFooter

JstUsername committed:
fix(NewsListWrapper): иправлены min-height для NewsListWrapper

JstUsername committed:
pr(ButtonLink): вместо ButtonLink использован /commons/RefreshButton

JstUsername committed:
pr(Suspense): исправлен иморт lazy, Suspense вынесен в HomePage

JstUsername committed:
pr($isVisible, $isVisibleIcon, $isExpand): явное указание типа в пропсах styled

JstUsername committed:
pr(calc): добавлены комментарии для значений в calc, скорректированы свойства для NewsListWrapper

JstUsername committed:
fix(NewsItemTitle.ts): вынесен в /commons/NewsItemTitle/

JstUsername committed:
pr(calc): calc вместе с комментариями вынесены в theme

JstUsername committed:
pr(store): store разделен на newsListState и newsItemState

JstUsername committed:
fix(newsLoading): удалено неиспользуемое поле newsLoading

JstUsername committed:
pr(newsItemState): лишние функции удалены из стейта

JstUsername committed:
pr(NewsItemPage): импорт NewsItemWrapper с помощью lazy, добавлен Suspense

JstUsername committed:
pr(CommentsItem): функция toggleExpandVisible вынесена из store в компонент

pr(CommentsItem): функция toggleExpandVisible вынесена из store в компонент

JstUsername committed:
pr(deleted, dead): добавлены проверки для deleted и dead комментариев

JstUsername committed:
pr(NewsCommentsBlock): добавлен локальный стейт и функция для добавления полей

JstUsername committed:
pr(toggleExpandVisible): функция исправлена и вынесена в NewsCommentsBlock

JstUsername committed:
pr(dependencies): обновлены зависимости для использования экспериментального use

pr(dependencies): обновлены зависимости для использования экспериментального use

JstUsername committed:
pr(CommentsItem): вместо рекурсивных функций использованы локальные стейты

JstUsername committed:
pr(NewsItemPage): удалены lazy и newsItem

JstUsername committed:
pr(HomePage): удалены lazy и проверка newsList.length

JstUsername committed:
pr(newsItemState, newsListState): функции запросов переписаны под возвращение промисов

JstUsername committed:
pr(newsList, newsItem): данные получены с помощью use

JstUsername committed:
fix(ts-ignore): обновлен tsconfig, удалены ts-ignore

JstUsername committed:
pr(newsListState): исправлены типы для data и newsList

JstUsername committed:
pr(extensions): добавлено правило import/extensions, исправлены импорты

JstUsername committed:
fix(NewsList.styled): удалены лишние комментарии

JstUsername committed:
pr(LinkToNews): вложенный 'a' вынесен в отдельный компонент

JstUsername committed:
pr(By): вложенный span вынесен в отдельный компонент

JstUsername committed:
pr(line breaks): добавлены разрывы строк между блоками кода

JstUsername committed:
pr(fetchNewsList): concat заменен на flat

JstUsername committed:
pr(newsItemState): из стейта удалены itemPageNotFound и resetNotFoundPage

JstUsername committed:
fix(ErrorHandler): исправлены пропы и функция сброса ошибки

JstUsername committed:
pr(NewsItemPage): установка ошибки 'Not found page' с помощью локального стейта

JstUsername committed:
fix(NewsItemPage): NewsContentBlock и NewsCommentsBlock разнесены по разным компонентам

JstUsername committed:
fix(App): исправлено отображение несуществующих путей

JstUsername committed:
fix(ErrorHandler): типы вынесены в ErrorHandler.types

Kimiyori commented on src/components/CommentsItem/CommentsItem.styled.ts:
```diff
@@ -0,0 +1,54 @@
+  ${({ theme }) => theme.typography.body1};
+  color: ${({ theme }) => theme.color.green};
+
+  p {
+    margin: 12px 0;
+    line-height: 1.2;
+    overflow: hidden;
+    text-overflow: ellipsis;
+  }
+
+  a {
+    color: ${({ theme }) => theme.color.green};
+    cursor: pointer;
+    white-space: nowrap;
+  }
+`;
```
убери вложенные стили

Kimiyori commented on src/components/CommentsItem/CommentsItem.styled.ts:
```diff
@@ -0,0 +1,54 @@
+`;
+
+export const CommentsChildItemWrapper = styled('div')<{ $isVisible: boolean | undefined }>`
+  display: ${({ $isVisible }) => ($isVisible ? 'block' : 'none')};
```
убери

Kimiyori commented on src/components/CommentsItem/CommentsItem.tsx:
```diff
@@ -0,0 +1,43 @@
+        </ExpandWrapper>
+        <CommentsItemContent dangerouslySetInnerHTML={{ __html: comment.content }} />
+      </div>
+      {comment.comments?.map((childComment: NewsItemType) => {
```
а это замени на это
 {isExpandVisible && comment.comments?.map((childComment: NewsItemType) => {
 мы же обговаривали это)

Kimiyori commented on src/components/ErrorHandler/ErrorHandler.tsx:
```diff
@@ -0,0 +1,22 @@
+          goHome()
+        </HomeLink>
+      ) : (
+        <RefreshButton onClick={() => window.location.reload()}>reloadPage()</RefreshButton>
```
вообще хорошо бы через navigate делать роутинг, а фукнцию возвращения передавать через проп

Kimiyori replied:
потому что по сообщению выдавать рендер ну такое себе) хотя бы поставь этот текст в консту какую-нибудь чотбы в нескольких местах ее не править если что

Kimiyori commented on src/components/NewsContentBlock/NewsContentBlock.tsx:
```diff
@@ -0,0 +1,44 @@
+  };
+
+  return (
+    newsItem !== null && (
```
просто проверяй на булевое значение с помощью !!

Kimiyori commented on src/pages/NewsItemPage/NewsItemPage.tsx:
```diff
@@ -0,0 +1,46 @@
+      getNewsItem(Number(id));
+      hasMounted.current = true;
+    }
+    setIsPageNotFound((prev) => !prev && true);
```
зачем это тут?

Kimiyori replied:
что мешает бросать ошибку в самой функции запроса?

IvanKoigerov replied:
А шо ничего не сделал? @JstUsername

Kimiyori commented on src/store/states/newsItemState/newsItemState.ts:
```diff
@@ -0,0 +1,30 @@
+  let data: NewsItemType | null = null;
+  const response = await fetch(`https://api.hnpwa.com/v0/item/${id}.json`);
+
+  if (response.status === 500) {
+    set({ itemServerDown: true });
+    return null;
```
вообще странно что только 500 обрабатываешь, есть и другие коды)

Kimiyori commented on src/components/NewsContentBlock/NewsContentBlock.tsx:
```diff
@@ -0,0 +1,44 @@
+import { NewsContentBlockProps } from './NewsContentBlock.types';
+
+export default function NewsContentBlock({ newsItem }: NewsContentBlockProps) {
+  const timestampToDate = (timestamp: number) => {
+    const date = fromUnixTime(timestamp);
+    return lightFormat(date, 'HH:mm dd.MM.yy').toString();
```
функцию можно вынести в какую-нибудь utils

Kimiyori commented on src/components/NewsContentBlock/NewsContentBlock.tsx:
```diff
@@ -0,0 +1,44 @@
+        <ContentBottomWrapper>
+          <Points>{newsItem.points} points</Points>
+          <ByTimeWrapper>
+            <By>
```
что за <By>? давай нормально называть!

Kimiyori commented on src/components/NewsItem/NewsItem.tsx:
```diff
@@ -0,0 +1,23 @@
+export default function NewsItem({ setIsPageNotFound }: NewsItemProps) {
+  const newsItem = use(useSelectorNewsItem());
+
+  useEffect(() => {
+    if (newsItem === null) {
+      setIsPageNotFound(true);
+    }
+  }, [setIsPageNotFound, newsItem]);
```
та же история, зачем тут это?

JstUsername committed:
pr(CommentsItem): удален CommentsChildItemWrapper, комментарии рендерятся по условию

JstUsername committed:
pr(errors): исправлены проброс и обработка ошибок

JstUsername committed:
pr(timestamp): функции для работы с timestamp вынесены в utils

JstUsername committed:
pr(Text): для span текста создан /commons/Text, удалены лишние styled-components

JstUsername committed:
fix(newsListState): удалена переменная data

JstUsername committed:
fix(ErrorBoundary): реализован сброс ошибки по нажатию на логотип.

Все компоненты в App обернуты в единый ErrorBoundary.

JstUsername committed:
fix(ErrorHandler): исправлена двойная отрисовка Header и Footer

IvanKoigerov commented on src/App.tsx:
```diff
@@ -0,0 +1,30 @@
+            <Route path="/" element={<Layout />}>
+              <Route index element={<HomePage />} />
+              <Route path="/news/:id" element={<NewsItemPage />} />
+              <Route path="*" element={<ErrorHandler error={new Error('Page not found')} clearLayout={true} />} />
```
`clearLayout` обычно пишем просто его

IvanKoigerov commented on src/App.tsx:
```diff
@@ -0,0 +1,30 @@
+    <BrowserRouter>
+      <ThemeProvider theme={theme}>
+        <GlobalStyle />
+        <ErrorBoundary FallbackComponent={ErrorHandler}>
```
У него вроде свой каллбэк есть у Reасt Router

IvanKoigerov commented on src/commons/Text/Text.ts:
```diff
@@ -0,0 +1,6 @@
+import styled from 'styled-components';
+
+export const Text = styled('span')<{ $variant: string; color: string }>`
+  ${({ theme, $variant }) => theme.typography[$variant]};
+  color: ${({ theme, color }) => theme.color[color]};
+`;
```
Ты не можешь ждать что вариант это какой-то там стринг, по логике это должно быть четкие значение, которые ты можешь ввести, то есть четкий тип. А то новый человек на проекте будет страдать и не понимать какие же есть варианты, да и тебе будет проще

IvanKoigerov commented on src/components/CommentsItem/CommentsItem.styled.ts:
```diff
@@ -0,0 +1,50 @@
+export const CommentsItemContent = styled('div')`
+  ${({ theme }) => theme.typography.body1};
+  color: ${({ theme }) => theme.color.green};
+
+  p {
+    margin: 12px 0;
+    line-height: 1.2;
+    overflow: hidden;
+    text-overflow: ellipsis;
+  }
+
+  a {
+    color: ${({ theme }) => theme.color.green};
+    cursor: pointer;
+    white-space: nowrap;
+  }
+`;
```
Не убрал

IvanKoigerov commented on src/components/CommentsItem/CommentsItem.styled.ts:
```diff
@@ -0,0 +1,50 @@
+  gap: 12px;
+`;
+
+export const ExpandIcon = styled(ArrowIcon)<{ $isVisibleIcon: boolean; $isExpand: boolean | undefined }>`
```
`$isExpand: boolean | undefined` почему именно такой? 
почему не $isExpand?: boolean
Тебе нужно чтобы разраб обязательно указал это значение?

IvanKoigerov commented on src/components/CommentsItem/CommentsItem.styled.ts:
```diff
@@ -0,0 +1,50 @@
+  gap: 12px;
+`;
+
+export const ExpandIcon = styled(ArrowIcon)<{ $isVisibleIcon: boolean; $isExpand: boolean | undefined }>`
+  width: 20px;
+  height: 16px;
+  fill: ${({ theme }) => theme.color.yellow};
+  display: ${({ $isVisibleIcon }) => ($isVisibleIcon ? 'block' : 'none')};
+  cursor: pointer;
+  transform: ${({ $isExpand }) => ($isExpand ? 'rotate(-180deg)' : 'rotate(0deg)')};
+`;
```
transition мб добавить?

IvanKoigerov commented on src/components/CommentsItem/CommentsItem.tsx:
```diff
@@ -0,0 +1,43 @@
+  const [isExpand, setIsExpand] = useState(false);
+
+  return (
+    <CommentsItemWrapper>
+      <div>
+        <ExpandWrapper>
+          <ExpandIcon
+            $isVisibleIcon={comment.comments.length !== 0}
+            $isExpand={isExpand}
+            onClick={() => {
+              setIsExpand(!isExpand);
+            }}
+          />
+          <CommentsItemUser>{comment.user}</CommentsItemUser>
+        </ExpandWrapper>
+        <CommentsItemContent dangerouslySetInnerHTML={{ __html: comment.content }} />
+      </div>
+      {isExpand &&
+        comment.comments?.map((childComment: NewsItemType) => {
+          if (childComment.deleted || childComment.dead) {
+            return null;
+          }
+          return (
+            <div key={childComment.id}>
+              <CommentsItem comment={childComment} />
+            </div>
+          );
+        })}
+    </CommentsItemWrapper>
+  );
```
старайся избегать лишних вещей в своей html, тут нет особого смысла использовать ещё один див будто бы. ни в ретерне ни в теле

IvanKoigerov commented on src/components/CommentsItem/CommentsItem.tsx:
```diff
@@ -0,0 +1,43 @@
+      </div>
+      {isExpand &&
+        comment.comments?.map((childComment: NewsItemType) => {
+          if (childComment.deleted || childComment.dead) {
+            return null;
+          }
+          return (
```
ыы пустой строчки тут нет

IvanKoigerov commented on src/theme.ts:
```diff
@@ -0,0 +1,48 @@
+    lg: '1440px',
+    xl: '1920px',
+  },
+  height: {
+    // @NOTE: 72px - высота header, 48px - высота footer (при ширине экрана ДО 1024px)
+    min: 'calc(100vh - (72px + 48px))',
+    // @NOTE: 88px - высота header, 64px - высота footer (при ширине экрана ОТ 1024px)
+    max: 'calc(100vh - (88px + 64px))',
+  },
```
такое себе название непонятно на месте что это за высота то, нужно исправлять и привязывать к сущности

IvanKoigerov commented on src/utils/timestampToAgo.ts:
обычно вся работа с датами лежит в одном файле, чтобы проще было работать, и в одном файле можно было найти все

IvanKoigerov commented on src/styles/GlobalStyle.ts:
```diff
@@ -0,0 +1,31 @@
+import { createGlobalStyle } from 'styled-components';
+
+export const GlobalStyle = createGlobalStyle`
+  body {
+    margin: 0;
+    background-color: ${({ theme }) => theme.color.backgroundDark};
+    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='400' height='400' viewBox='0 0 800 800'%3E%3Cg fill='none' stroke='%23022741' stroke-width='1'%3E%3Cpath d='M769 229L1037 260.9M927 880L731 737 520 660 309 538 40 599 295 764 126.5 879.5 40 599-197 493 102 382-31 229 126.5 79.5-69-63'/%3E%3Cpath d='M-31 229L237 261 390 382 603 493 308.5 537.5 101.5 381. ...
```
вешаешь стили на body, я уже ругал тебя за это, этого делать нельзя, не на всех страницах ты можешь захотеть видеть это

JstUsername committed:
pr(dateUtils): функции для работы с датами вынесены в один файл, исправлены импорты

JstUsername committed:
pr(CommentsItem): удален лишний обертывающий div

JstUsername committed:
pr(background): фон вынесен из глобальных стилей в AppWrapper

JstUsername committed:
pr(App): исправлена передача clearLayout пропа

JstUsername committed:
pr(Text): исправлена типизация пропов $variant: и color

JstUsername committed:
pr(CommentsItem): добавлен разрыв после if, добавлен transition

JstUsername committed:
pr(theme): исправлен нейминг для theme.height
